src/leap/common/events/client.py

```diff
@@ -55,275 +55,276 @@ class CallbackAlreadyRegistered(Exception):
+        if not uid:
+            uid = uuid.uuid4()
+        elif uid in self._callbacks[event] and not replace:
+            raise CallbackAlreadyRegistered()
+        self._callbacks[event].append((uid, callback))
+        self._sub.socket.setsockopt(zmq.SUBSCRIBE, str(event))
+        return uid
 
-        :param port: the port in which the daemon should listen
-        :type port: int
+    def unregister(self, event, uid=None):
+        """
+        Unregister callbacks for an event.
+
+        If uid is not None, then only the callback identified by the given uid
+        is removed. Otherwise, all callbacks for the event are removed.
 
-        :return: a daemon instance
-        :rtype: EventsClientDaemon
+        :param event: The event that triggers the callback.
+        :type event: Event
+        :param uid: The callback uid.
+        :type uid: str
         """
-        return cls.ensure_service(port, EventsClientService())
+        self.ensure_client()
+        if not uid:
+            logger.debug(
+                "Unregistering all callbacks from event %s." % event)
+            self._callbacks[event] = {}
+        else:
+            logger.debug(
+                "Unregistering callback %s from event %s." % (uid, event))
+            self._callbacks[event] = filter(
+                lambda u, _: u == uid,
+                self._callbacks[event])
+        if not self._callbacks[event]:
+            del self._callbacks[event]
+            self._sub.socket.setsockopt(zmq.UNSUBSCRIBE, str(event))
+
+    def emit(self, event, *content):
+        """
+        Send an event.
+
+        :param event: The event to be sent.
+        :type event: Event
+        :param content: The content of the event.
+        :type content: list
+        """
+        self.ensure_client()
+        logger.debug("Sending event: (%s, %s)" % (event, content))
+        # add send() as a callback for ioloop so it works between threads
+        self._loop.add_callback(
+            lambda: self._push.send(
+                str(event) + b'\0' + pickle.dumps(content)))
+
+    def run(self):
+        """
+        Run the events client.
+        """
+        logger.debug("Starting ioloop.")
+        self._init_zmq()
+        self._initialized.set()
+        self._loop.start()
+        self._loop.close()
+        logger.debug("Ioloop finished.")
+
+    def ensure_client(self):
+        """
+        Make sure the events client thread is started.
+        """
+        with self._lock:
+            if not self.is_alive():
+                self.start()
+                self._initialized.wait()
+
+    def shutdown(self):
+        """
+        Shutdown the events client thread.
+        """
+        logger.debug("Shutting down client...")
+        with self._lock:
+            if self.is_alive():
+                self._loop.stop(wait=True)
+
+
+def shutdown():
+    """
+    Shutdown the events client thread.
+    """
+    EventsClient.instance().shutdown()
+
```

Comment:
If automatically shutting down is a concern, I'd give a try at atexit.register() https://docs.python.org/2/library/atexit.html